src/utils/function.js

```diff
@@ -44,11 +44,50 @@ export function rateLimit(win, callback, minInterval) {
     }
   }
 
-  return function() {
+  return function(...args) {
     if (locker) {
-      nextCallArgs = arguments;
+      nextCallArgs = args;
     } else {
-      fire(arguments);
+      fire(args);
+    }
+  };
+}
+
+/**
+ * Wraps a given callback and applies a wait timer, so that minInterval
+ * milliseconds must pass since the last call before the callback is actually
+ * invoked.
+ *
+ * @param {!Window} win
+ * @param {function()} callback
+ * @param {number} minInterval the minimum time interval in millisecond
+ * @returns {function()}
+ */
+export function debounce(win, callback, minInterval) {
+  let locker = 0;
+  let timestamp = 0;
+  let nextCallArgs = null;
+
+  function fire(args) {
+    nextCallArgs = null;
+    callback.apply(null, args);
+  }
+
+  function waiter() {
+    locker = 0;
+    const remaining = minInterval - (Date.now() - timestamp);
+    if (remaining > 0) {
+      locker = win.setTimeout(waiter, remaining);
+    } else {
+      fire(nextCallArgs);
+    }
+  }
+
+  return function(...args) {
+    timestamp = Date.now();
```

Comment:
Soo mannyyy commentttts.

> Is there a better common name/API for debounce out there? 

`debounce` is the standard name, since Underscore released it some 6 years ago.

> what is even a clean name for variable that holds the return of debounce?

`callback` => `debouncedCallback`, same as we use `callback` => `boundCallback`

> also a bit fragile (e.g. replacing debouncedScroll(); with debounce(onEnd, 500); will break the code).

That's like replacing `boundMethod()` with `method.bind(this)`. One creates a wrapped function, the other executes it.

> Something like `listenerTimeout(onScroll, 500).then(onEnd);`

You're looking for an `Observable`, specifically a [debounced one](http://reactivex.io/rxjs/class/es6/Observable.js~Observable.html#instance-method-debounce).

> Maybe correct answer is to fix the first "it is fragile" problem by using two listeners:

That's basically what I was suggesting in the `PositionObserver` PR.